src/m1_facilitator/handlers/GatewayProvenHandler.ts

```diff
@@ -0,0 +1,111 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as Web3Utils from 'web3-utils';
+import assert from 'assert';
+import BigNumber from 'bignumber.js';
+
+import Gateway from '../models/Gateway';
+import GatewayRepository, { LastProvenBlockNumberIsNotStrictlyGrowingError } from '../repositories/GatewayRepository';
+import Logger from '../../common/Logger';
+
+
+/**
+ * GatewayProvenEntityInterface represents a gateway-proven subgraph entity.
+ */
+export interface GatewayProvenEntityInterface {
+  contractAddress: string;
+  remoteGateway: string;
+  blockNumber: string;
+}
+
+/**
+ * GatewayProven class wraps a gateway-proven subgraph entity
+ * for type checks and conversions.
+ */
+class GatewayProven {
+  public readonly gatewayAddress: string;
+
+  public readonly remoteGatewayAddress: string;
+
+  public readonly blockNumber: BigNumber;
+
+  public constructor(entity: GatewayProvenEntityInterface) {
+    assert(Web3Utils.isAddress(entity.contractAddress));
+    assert(Web3Utils.isAddress(entity.remoteGateway));
+    assert(entity.blockNumber !== '');
+
+    this.gatewayAddress = Web3Utils.toChecksumAddress(entity.contractAddress);
+    this.remoteGatewayAddress = Web3Utils.toChecksumAddress(entity.remoteGateway);
+    this.blockNumber = new BigNumber(entity.blockNumber);
+  }
+}
+
+/** GatewayProvenHandler class handles gateway-proven subgraph entities. */
+export default class GatewayProvenHandler {
+  private readonly gatewayRepository: GatewayRepository;
+
+  /** Constructs GatewayProvenHandler from the specified arguments. */
+  public constructor(gatewayRepository: GatewayRepository) {
+    this.gatewayRepository = gatewayRepository;
+  }
+
+  /**
+   * handle() function accepts an array subgraph gateway-proven entities.
+   * If there is no model in the gateway repository matching to a gateway
+   * global address of an entity, the function logs a warning without failing.
+   * Otherwise, it updates a model in the gateway repository matchin to a
+   * gateway global address of an entity with the new block number (of a gateway
+   * proof). The gateway repository will throw an exception if the newly
+   * proposed block number is less than or equal to the stored one.
+   * The function catches the exception and logs a warning without failing.
+   */
+  public async handle(entities: GatewayProvenEntityInterface[]): Promise<void> {
+    const savePromises = entities.map(async (entity): Promise<void> => {
+      const gatewayProven = new GatewayProven(entity);
```

Comment:
Why is the significance of the `GatewayProven` class?
We already have `GatewayProvenEntityInterface`

Reply:
GatewayProvenEntityInterface represents subgraph entity (like, blockNumber type is `string`). However, GatewayProven class does type checks and conversion of that entity.